Issue #248: Making custom, specific exceptions

nikkisquared opened the pull request:
Note that this pull request is not ready for merging, but I'm opening it now to ask for feedback. I started working on issue #248 as suggested by adusca. So far I've spent most of my time trying to figure out the level of specificity I should be making. For some edits I feel like it's just right, but others feel too specific. I'd really appreciate any feedback on what I'm doing right/wrong, and what I can do to improve my changes. Thanks!

armenzg commented on mozci/mozci.py:
```diff
@@ -40,6 +40,14 @@
 QUERY_SOURCE = BuildApi()
 
 
+class JobException(Exception):
```
For the record, I have no clear view of what right should look like.
Let's pick something that it more or less feels right :)

Can you change these two exceptions to be MozciException? It makes implicit reference to the module this code lives within (mozci.py) and it is not as specific to the actual issue raised.

armenzg commented on mozci/platforms.py:
```diff
@@ -9,6 +9,10 @@
 LOG = logging.getLogger('mozci')
 
 
+class BuildException(Exception):
```
Could you import the MozciException from the other module?
platforms.py was kind of an extension of mozci.py

vaibhavmagarwal replied:
This will cause a circular import since platforms.py is already imported in mozci.py
We can define the MozciException here once again or let this remain as a normal base exception.

vaibhavmagarwal replied:
Or we can introduce another file "errors.py" and define MozciException there and import it in all the places.

armenzg commented on mozci/scripts/generate_triggercli.py:
```diff
@@ -44,6 +44,14 @@
+    pass
+
+
+class BugCommentException(Exception):
```
Let's ignore all scripts completely.
All files under mozci/scripts/ are not components that external consumers will be using.
The scripts/ directory lives under mozci instead of the root directory because the usage we have of console_scripts under setup.py which requires us to put it under mozci/
One day I hope we will get rid of that.

armenzg commented on mozci/scripts/trigger.py:
```diff
@@ -13,6 +13,10 @@
 from mozci.sources.pushlog import query_repo_tip
 
 
+class ValidationException(Exception):
```
Let's ignore this script as well.

armenzg commented on mozci/sources/buildapi.py:
```diff
@@ -31,6 +31,10 @@ class BuildapiException(Exception):
     pass
 
 
+class RepoNameException(Exception):
```
Let's call it BuildApiException.

armenzg replied:
I realized we already have the BuildapiException inside of this module. Can you please use that?

armenzg commented on mozci/sources/pushlog.py:
```diff
@@ -27,6 +27,10 @@
 VALID_CACHE = {}
 
 
+class PushlogException(Exception):
```
Good :+1:

armenzg commented on mozci/utils/transfer.py:
```diff
@@ -26,6 +26,10 @@
 MEMORY_SAVING_MODE = False
 
 
+class StatusException(Exception):
```
Let's import the MozciException.

Hrmm I never realized how complicated naming exceptions is.

armenzg commented:
Hi @nikkisquared, thank you for this patch!

Could you also try to add information about the exceptions in the docstrings where it applies?
In the pep 257 docs it is suggested [1]
My apologies about the modules not being properly documented according to that standard but I would like to add info about the exceptions so consumers can be aware of it (even if the rest of the docstring is not complete).

[1] https://www.python.org/dev/peps/pep-0257/
The docstring for a function or method should summarize its behavior and document its arguments, return value(s), side effects, exceptions raised, and restrictions on when it can be called (all if applicable)

armenzg commented:
Having mozci/exceptions.py would be useful.

nikkisquared commented:
I've just pushed my second draft, following all of your suggestions for the names and docstring usage. Thanks a lot for the advice @armenzg, you answered pretty much every question I had in mind! The next thing for me to do would be writing documentation for the exceptions I haven't done anything to, and possibly putting all the custom exceptions in mozci/exceptions.py.

There's two exceptions I haven't done anything with, which are the "_valid_builder" function in mozci/sources/buildapi.py that instantly raises an error when called, and the exception in "_load_json_file" mozci/utils/transfer.py that only comes up if a Windows user doesn't have gzip installed. What should I do about either of those, if anything?

armenzg commented:
For [1], it perhaps make sense to simply say "pass" and turn the exceptions message to be a simple comment. We're simply not to use that method. We could even remove it completely.
The exception perhaps was unneccesarily dramatic :)

For [2], I doubt we will have tools being built using Windows that would require us to promise them a known exception.

For reference:
[1] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/sources/buildapi.py#L156
[2] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/utils/transfer.py#L104

armenzg commented:
The PR looks good to me.

WRT to "@ adusca armenzg vaibhav1994 I think I should put "Error" instead of "Exception" in each of my custom exceptions, following what Python and some existing code does"

Do you have a reference to using Error instead?
I'm happy either way but I would like to know what you're refering to.

armenzg commented:
@chmanchester since you requested Mozci to have more formalized exceptions (rather than generic ones), would you mind having a look to see if this meets your expectations? Thanks!

nikkisquared commented:
@armenzg I should've linked what I was looking at in my last comment, sorry. The standard python exceptions all use Error in the name https://docs.python.org/2/c-api/exceptions.html#standard-exceptions and I noticed that sources/buildapi.py imports AuthenticationError from another mozci module. So it seems like changing Exception to Error would be good practice. Also @vaibhavmagarwal (whose comments here I can't seem to find again) originally suggested the Exceptions.py module be called Errors.py, and I hadn't thought about why that might be until now.

nikkisquared commented:
I've renamed all the exceptions to use Error now, and removed the the "_valid_builder" function.

I made sure my work was done with `grep "(Exception) * -r` and found that there are three custom-named Exceptions in the source code [1] [2] [3] that I could put in errors.py and import from there. I could also re-name the latter two to be "Error" instead of "Exception. Thoughts @armenzg?

[1] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/query_jobs.py#L17
[2] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/sources/buildjson.py#L22
[3] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/utils/authentication.py#L20

armenzg commented:
Thanks for clarifying about the error naming. It makes sense.

Please go ahead and move those exceptions into the errors module and rename them properly.
We can merge after that.

nikkibee: would you be interested on contributing in other issues as well?
Also, would you mind making a mozillians.org profile once you have a chance? It's easier to help Mozilla reach out to you. I can vouch for you.

armenzg commented:
Hi Nikkibee,
Could you also try to squash your commits?
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

I would like to merge one commit instead of multiple ones.
I have not done this myself but I have been aware that it is a good practice which I would like to get started.
Let me know if you have trouble.

nikkisquared commented:
@armenzg I'll move the other three exceptions into errors.py and then see about squashing the commits. Then I'll make a mozillians account. Thanks again for all the feedback!

chmanchester commented:
These exceptions look good to me, thanks for the patch!

nikkisquared commented:
Your welcome @chmanchester! I've enjoyed working on it.
